client/write.go

```diff
@@ -131,31 +136,29 @@ func (w *writeState) close() {
 }
 
 func (w *writeState) completionFn(result interface{}, err error) {
-	remaining := atomic.AddInt32(&w.pending, -1)
-	successful := int32(0)
-	if err == nil {
-		successful = atomic.AddInt32(&w.successful, 1)
-	}
+	hostID := result.(topology.Host).ID()
+	// NB(bl) panic on invalid result, it indicates a bug in the code
 
 	w.Lock()
+	w.pending--
 
 	if err != nil {
 		w.errors = append(w.errors, err)
+	} else if hostShardSet, ok := w.topoMap.LookupHostShardSet(hostID); !ok {
+		w.errors = append(w.errors, fmt.Errorf("Missing host shard in writeState completionFn: %s", hostID))
+	} else if shardState, err :=
+		hostShardSet.ShardSet().LookupStateByID(w.op.shardID); err != nil {
+		w.errors = append(w.errors, fmt.Errorf("Missing shard %d in host %s", w.op.shardID, hostID))
```

Comment:
To stay consistent with our other errors make these both start with lower case characters yeah?